File: backend/src/python/adapters/coo.ts
/**
 * coo adapter — deterministic pre-flight + delivery checks.
 *
 * Ported from python-services/src/financex/calculators/coo.py so the
 * rules run in-process without a Python subprocess. Keep the two
 * sides in sync — if you add a _PreflightRule in Python, mirror it
 * in PREFLIGHT_RULES below (and vice-versa).
 *
 * The LLM coo agent does more than this (directive issuance, memory
 * lookup, qualitative checks). When PYTHON_COO_ENABLED is off the
 * LLM path keeps running — this module is a rollback target for the
 * deterministic subset.
 */

export type Sector = 'industrial' | 'banking' | 'holding' | 'insurance' | 'reit';
export type PreflightDecision = 'go' | 'conditional' | 'no_go';
export type DeliveryDecision = 'approved' | 'revision_needed' | 'blocked';


export interface CheckItem {
  code: string;
  label: string;
  passed: boolean;
  message: string;
}


export interface PreflightReport {
  ticker: string;
  sector: Sector;
  decision: PreflightDecision;
  items: CheckItem[];
}


export interface DeliveryReport {
  ticker: string;
  decision: DeliveryDecision;
  items: CheckItem[];
}


interface PreflightRule {
  code: string;
  label: string;
  sectors: ReadonlyArray<Sector>;
}


const ALL_SECTORS: ReadonlyArray<Sector> = ['industrial', 'banking', 'holding', 'insurance', 'reit'];

const PREFLIGHT_RULES: ReadonlyArray<PreflightRule> = [
  { code: 'KAP_ACCESS',          label: 'Reachable KAP endpoint (api/search/combined)', sectors: ALL_SECTORS },
  { code: 'TCMB_FX',             label: 'TCMB daily FX bulletin reachable',             sectors: ALL_SECTORS },
  { code: 'FIVE_YEAR_WINDOW',    label: 'Five-year fetch window resolvable',             sectors: ALL_SECTORS },
  { code: 'BDDK_FORMAT_AWARE',   label: 'Parser loaded banking label overlay',           sectors: ['banking'] },
  { code: 'HOLDING_SOTP_NOTED',  label: 'Holding SOTP requirement flagged to valuation', sectors: ['holding'] },
];


export function runPreflight(
  ticker: string,
  sector: Sector,
  facts: Record<string, boolean> = {},
): PreflightReport {
  const items: CheckItem[] = [];
  let failedAny = false;

  for (const rule of PREFLIGHT_RULES) {
    if (!rule.sectors.includes(sector)) continue;
    // Optimistic default: missing probe = assume OK so preflight never
    // blocks just because a liveness probe wasn't run.
    const passed = facts[rule.code] !== false;
    if (!passed) failedAny = true;
    items.push({
      code: rule.code,
      label: rule.label,
      passed,
      message: passed ? 'ok' : `fact missing or false for ${rule.code}`,
    });
  }

  return {
    ticker: ticker.toUpperCase(),
    sector,
    decision: failedAny ? 'no_go' : 'go',
    items,
  };
}


const TABLE_OPEN_RE = /<table\b/gi;
const TABLE_CLOSE_RE = /<\/table>/gi;


/**
 * Phase F (2026-04-28) — QA gate context for delivery. When the QA truth
 * gate produced a hard_fail or block_publish escalation, delivery
 * decision must be 'blocked' regardless of HTML structural checks. The
 * board reader sees the publish-block banner in the report itself; this
 * gate ensures the COO/orchestrator agrees and refuses to mark the
 * report as approved.
 */
export interface DeliveryQaContext {
  qa_decision?: string;
  escalation_recommendation?: string;
  blocker_failures?: string[];
  // Phase D — when target_price publish was blocked at the valuation
  // layer (holding without curated SOTP YAML), delivery should also
  // refuse 'approved' until the operator curates the SOTP file.
  target_price_publish_blocked?: boolean;
}

export function runDeliveryCheck(
  ticker: string,
  html: string,
  qa?: DeliveryQaContext,
): DeliveryReport {
  const items: CheckItem[] = [];
  const lower = html.toLowerCase();

  // 1. HTML structural integrity
  const openTables = (html.match(TABLE_OPEN_RE) ?? []).length;
  const closeTables = (html.match(TABLE_CLOSE_RE) ?? []).length;
  const tablesOk = openTables === closeTables;
  items.push({
    code: 'TABLE_BALANCE',
    label: 'Every <table> closes properly',
    passed: tablesOk,
    message: `${openTables} opened, ${closeTables} closed`,
  });

  // 2. Document envelope
  const envelopeOk = lower.includes('<html') && lower.includes('</html>');
  items.push({
    code: 'HTML_ENVELOPE',
    label: 'HTML envelope present',
    passed: envelopeOk,
    message: envelopeOk ? 'ok' : 'missing <html> or </html>',
  });

  // 3. Mandatory SPK disclaimer
  const disclaimerOk = lower.includes('yatırım tavsiyesi değildir');
  items.push({
    code: 'SPK_DISCLAIMER',
    label: 'Mandatory SPK disclaimer present',
    passed: disclaimerOk,
    message: disclaimerOk ? 'ok' : "missing 'yatırım tavsiyesi değildir'",
  });

  // 4. Minimum page-count proxy — full LLM template is 80 KB+; minimal
  // Python template is ~6 KB. Floor at 5 KB to catch stubs / errors.
  const sizeOk = html.length >= 5_000;
  items.push({
    code: 'MIN_PAYLOAD_SIZE',
    label: 'Rendered HTML ≥ 5 KB',
    passed: sizeOk,
    message: `${html.length} bytes`,
  });

  // 5. Phase F — QA gate consistency. If QA produced hard_fail or
  // escalation block_publish, delivery MUST not approve.
  const qaDecision = String(qa?.qa_decision ?? '');
  const qaEscalation = String(qa?.escalation_recommendation ?? '');
  const qaHardFail = qaDecision === 'hard_fail' || qaEscalation === 'block_publish';
  const qaGateOk = !qaHardFail;
  items.push({
    code: 'QA_GATE_PASS',
    label: 'QA truth gate not hard-failed',
    passed: qaGateOk,
    message: qaGateOk
      ? 'qa_decision passes (not hard_fail / block_publish)'
      : `QA hard-fail (qa_decision=${qaDecision || 'n/a'}, escalation=${qaEscalation || 'n/a'}; blockers=${(qa?.blocker_failures ?? []).join(',') || 'n/a'})`,
  });

  // 6. Phase D — SOTP / target_price publish block. Holding without
  // curated SOTP YAML must not deliver a publishable target price.
  const sotpOk = !qa?.target_price_publish_blocked;
  items.push({
    code: 'SOTP_PUBLISH_GATE',
    label: 'SOTP gate / target_price publishable',
    passed: sotpOk,
    message: sotpOk
      ? 'target_price publish allowed (or non-holding)'
      : 'target_price publish blocked (holding without curated SOTP)',
  });

  let decision: DeliveryDecision;
  if (!envelopeOk || !disclaimerOk || !qaGateOk) decision = 'blocked';
  else if (!tablesOk || !sizeOk || !sotpOk) decision = 'revision_needed';
  else decision = 'approved';

  return {
    ticker: ticker.toUpperCase(),
    decision,
    items,
  };
}


// ---------------------------------------------------------------------
// Phase dispatcher — the orchestrator tells us which phase we're in
// (preflight before data_collection, delivery after report_formatter)
// ---------------------------------------------------------------------

export type CooPhase = 'preflight' | 'delivery';


export interface LegacyCooOutput {
  agent_id: string;
  output_id: string;
  ticker: string;
  phase: CooPhase;
  decision: PreflightDecision | DeliveryDecision;
  checks: CheckItem[];
  directives: string[];
  warnings: string[];
  review_status: string;
  source: 'python';
}


export function adaptCooForLegacy(
  report: PreflightReport | DeliveryReport,
  phase: CooPhase,
  outputId: string,
): LegacyCooOutput {
  const warnings = report.items
    .filter(it => !it.passed)
    .map(it => `${it.code}: ${it.message}`);

  const directives: string[] = [];
  if (phase === 'preflight') {
    const pre = report as PreflightReport;
    if (pre.sector === 'banking') directives.push('Banking filer — BDDK label overlay + NII_RECONCILE rules apply.');
    if (pre.sector === 'holding') directives.push('Holding — SOTP valuation required; consolidated DCF is upper-bound only.');
  }

  return {
    agent_id: 'coo',
    output_id: outputId,
    ticker: report.ticker,
    phase,
    decision: report.decision,
    checks: report.items,
    directives,
    warnings,
    review_status: 'pending_ceo_review',
    source: 'python',
  };
}

File: python-services/src/financex/calculators/coo.py
"""coo — pre-flight and delivery checks, purely rule-based.

Two phases of the pipeline use this module:

  - pre-flight (before data_collection): sector-aware mandatory
    requirements (which KAP disclosures, which external data sources,
    which reconciliation checks apply).
  - delivery (after report_formatter): HTML bütünlüğü, tablo kapanışı,
    mandatory disclaimer, page-count sanity.

Both phases return a structured decision so the orchestrator can
either proceed, proceed-with-flags, or block.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import UTC, datetime
from decimal import Decimal

from pydantic import Field

from financex.schemas.base import FinancexModel, Sector


class CheckItem(FinancexModel):
    code: str
    label: str
    passed: bool
    message: str


class PreflightReport(FinancexModel):
    ticker: str
    sector: Sector
    decision: str          # 'go' | 'conditional' | 'no_go'
    items: list[CheckItem] = Field(default_factory=list)
    computed_at: datetime
    schema_version: str = "1.0.0"


class DeliveryReport(FinancexModel):
    ticker: str
    decision: str          # 'approved' | 'revision_needed' | 'blocked'
    items: list[CheckItem] = Field(default_factory=list)
    computed_at: datetime
    schema_version: str = "1.0.0"


# ---------- Pre-flight catalogue ----------

@dataclass(frozen=True)
class _PreflightRule:
    code: str
    label: str
    sectors: tuple[Sector, ...]


_PREFLIGHT_RULES: tuple[_PreflightRule, ...] = (
    _PreflightRule(
        "KAP_ACCESS", "Reachable KAP endpoint (api/search/combined)",
        sectors=tuple(Sector)
    ),
    _PreflightRule(
        "TCMB_FX", "TCMB daily FX bulletin reachable",
        sectors=tuple(Sector)
    ),
    _PreflightRule(
        "FIVE_YEAR_WINDOW", "Five-year fetch window resolvable",
        sectors=tuple(Sector)
    ),
    _PreflightRule(
        "BDDK_FORMAT_AWARE", "Parser loaded banking label overlay",
        sectors=(Sector.BANKING,)
    ),
    _PreflightRule(
        "HOLDING_SOTP_NOTED", "Holding SOTP requirement flagged to valuation",
        sectors=(Sector.HOLDING,)
    ),
)


def preflight(
    ticker: str,
    sector: Sector,
    *,
    facts: dict[str, bool] | None = None,
) -> PreflightReport:
    """Run sector-filtered pre-flight checks.

    `facts` supplies live probe results — e.g. {"KAP_ACCESS": True}.
    Missing keys default to True (optimistic) so pre-flight doesn't
    require every probe before the pipeline can even run.
    """
    facts = facts or {}
    items: list[CheckItem] = []
    failed_any = False
    for rule in _PREFLIGHT_RULES:
        if sector not in rule.sectors:
            continue
        passed = facts.get(rule.code, True)
        if not passed:
            failed_any = True
        items.append(CheckItem(
            code=rule.code,
            label=rule.label,
            passed=passed,
            message="ok" if passed else f"fact missing or False for {rule.code}",
        ))

    decision = "no_go" if failed_any else "go"
    return PreflightReport(
        ticker=ticker.upper(),
        sector=sector,
        decision=decision,
        items=items,
        computed_at=datetime.now(UTC),
    )


# ---------- Delivery checks (HTML scan) ----------

_CRITICAL_STRINGS = (
    "<html",
    "</html>",
    "yatırım tavsiyesi",   # mandatory disclaimer
)

_TABLE_OPEN_RE = re.compile(r"<table\b", re.IGNORECASE)
_TABLE_CLOSE_RE = re.compile(r"</table>", re.IGNORECASE)


def delivery_check(ticker: str, html: str) -> DeliveryReport:
    """Scan a rendered HTML report for structural + compliance requirements."""
    items: list[CheckItem] = []

    # 1. HTML structural integrity
    open_tables = len(_TABLE_OPEN_RE.findall(html))
    close_tables = len(_TABLE_CLOSE_RE.findall(html))
    tables_ok = open_tables == close_tables
    items.append(CheckItem(
        code="TABLE_BALANCE",
        label="Every <table> closes properly",
        passed=tables_ok,
        message=f"{open_tables} opened, {close_tables} closed",
    ))

    # 2. Document envelope
    envelope_ok = "<html" in html.lower() and "</html>" in html.lower()
    items.append(CheckItem(
        code="HTML_ENVELOPE",
        label="HTML envelope present",
        passed=envelope_ok,
        message="ok" if envelope_ok else "missing <html> or </html>",
    ))

    # 3. Mandatory SPK disclaimer (paraphrase match — must appear)
    disclaimer_ok = "yatırım tavsiyesi değildir" in html.lower()
    items.append(CheckItem(
        code="SPK_DISCLAIMER",
        label="Mandatory SPK disclaimer present",
        passed=disclaimer_ok,
        message="ok" if disclaimer_ok else "missing 'yatırım tavsiyesi değildir'",
    ))

    # 4. Minimum page-count proxy — the minimal Python template runs
    #    ~6 KB; the full LLM-authored template is 80 KB+. We floor at
    #    5 KB: anything smaller is clearly a stub / error placeholder.
    size_ok = len(html) >= 5_000
    items.append(CheckItem(
        code="MIN_PAYLOAD_SIZE",
        label="Rendered HTML ≥ 5 KB",
        passed=size_ok,
        message=f"{len(html)} bytes",
    ))

    # Decide
    if not envelope_ok or not disclaimer_ok:
        decision = "blocked"
    elif not tables_ok or not size_ok:
        decision = "revision_needed"
    else:
        decision = "approved"

    return DeliveryReport(
        ticker=ticker.upper(),
        decision=decision,
        items=items,
        computed_at=datetime.now(UTC),
    )
